### PR TITLE
11 finish events rewrite

### LINKED ISSUE #11: Finish Events feature by implementing Create, Update, and Delete API (and GUI) functionality

### PR DESCRIPTION
**Pull request by zyphlar**
Fixes #11
Includes #32

### CODE REVIEW

**Comment by duncanbeevers** [comment]
Here's my diff where I clear the db after each experiment. Seems to get things passing (except for the DELETE event test, which I think is still broken on purpose)

```diff
e8435d5 Duncan B.. 82 secon.. fixup! get events tests passing, prep app for heroku

diff --git a/docker-run.sh b/docker-run.sh
index 5766867..1003938 100755
--- a/docker-run.sh
+++ b/docker-run.sh
@@ -5,25 +5,25 @@ echo "Setting PATH from $PATH ..."
 export PATH=$PATH:$(pwd)/node_modules/.bin
 echo "... to $PATH"
 
-if [[ $PORT  ]]
+if [ $PORT  ]
 then
   export NODE_PORT=$PORT
   echo "Setting NODE_PORT=$PORT based on PORT"
 fi
 
-if [[ $NPMINSTALL == 1 ]]
+if [ $NPMINSTALL == 1 ]
 then
   npm install
 fi
 
-if [[ $NODE_ENV == "production" ]]
+if [ $NODE_ENV == "production" ]
 then
   npm run start
-elif [[ $NODE_ENV == "development" ]]
+elif [ $NODE_ENV == "development" ]
 then
   npm run db_migrate_latest
   npm run develop
-elif [[ $NODE_ENV == "test" ]]
+elif [ $NODE_ENV == "test" ]
 then
   npm run db_migrate_rollback_all
   npm run db_migrate_latest
diff --git a/test/fixture-client.js b/test/fixture-client.js
index 99e8b02..7e53a2d 100644
--- a/test/fixture-client.js
+++ b/test/fixture-client.js
@@ -17,10 +17,10 @@ const { forEach, keys, values } = require('lodash');
 const promise = require('bluebird');
 const debug = require('debug')('errors');
 const url = require('url');
+const { Factory } = require('rosie');
 const server = require('..');
 const { connection } = require('../knexfile');
 const knex = require('../knex');
-const { Factory } = require('rosie');
 
 const fixturesConfig = {
   client: 'pg',
@@ -44,7 +44,7 @@ const tableToDeleteKey = {
 };
 
 async function mapRelation(relationValue) {
-  console.warn("mapRelation:",relationValue);
+  console.warn('mapRelation:', relationValue);
   const [tableName, offset] = relationValue.split(':');
   return (await knex(tableName).offset(offset).first('id')).id;
 }
@@ -54,7 +54,7 @@ async function mapRelation(relationValue) {
 // collection of untransformed fixtures and replaces the `created_by` field
 // of each one with the corresponding user id from the database.
 function createMapRelations(relationNames) {
-  console.warn("Mapping relation:",relationNames);
+  console.warn('Mapping relation:', relationNames);
   return async function mapRelations(collection) {
     return Promise.all(collection.map(async (item) => ({
       ...item,
@@ -89,8 +89,6 @@ module.exports = {
       const ids = values(table)[0];
       const key = tableToDeleteKey[tableName];
 
-      console.log("deleting", tableName);
-
       return knex(tableName)
         .whereIn(key, ids)
         .del()
@@ -102,11 +100,17 @@ module.exports = {
     });
   },
   async makeUserIdAdmin(userId) {
+    await knex('groups')
+      .insert({ id: 'ADMIN', description: 'Admin users' })
+      .onConflict('id')
+      .merge();
+
     const membership = new Factory();
     membership
       .attr('user_id')
       .attr('group_id');
-    memberships = [
+
+    const memberships = [
       // user 1 will be an admin
       membership.build({
         user_id: userId,
diff --git a/test/routes/events/add.js b/test/routes/events/add.js
index 9fd9f43..4f5f431 100644
--- a/test/routes/events/add.js
+++ b/test/routes/events/add.js
@@ -18,19 +18,18 @@ const lab = exports.lab = require('lab').script();
 const url = require('url');
 
 const server = require('../../..');
-const { destroyRecords, getAuthToken, fixtures, makeUserIdAdmin } = require('../../fixture-client');
+const { getAuthToken, makeUserIdAdmin } = require('../../fixture-client');
 const { users } = require('../../fixtures');
 const knex = require('../../../knex');
+const clearDb = require('../../clearDb');
 
 lab.experiment('POST /events', () => {
   let Authorization;
   let myUserId;
-  let memberships;
-  let events;
 
   lab.before(async () => {
-    let insertedUserIds = await knex('users').insert(users).returning(['id']);
-    myUserId = insertedUserIds[0]['id'];
+    const insertedUserIds = await knex('users').insert(users).returning(['id']);
+    myUserId = insertedUserIds[0].id;
 
     await makeUserIdAdmin(myUserId);
 
@@ -38,8 +37,8 @@ lab.experiment('POST /events', () => {
     Authorization = authRes.token;
   });
 
-  lab.after(() => {
-    return destroyRecords({ memberships }).then(destroyRecords({ events })).then(destroyRecords({ users }));
+  lab.after(async () => {
+    await clearDb();
   });
 
   lab.test('should create an event with appropriate created_by', async () => {
@@ -74,11 +73,11 @@ lab.experiment('POST /events', () => {
     const res2 = await server.inject(options2);
     expect(res2.statusCode).to.equal(200);
     expect(res2.result).to.be.an.array();
-    expect(res2.result[res2.result.length-1].name).to.equal('Laser Class Testy');
-    expect(res2.result[res2.result.length-1].created_by).to.equal(myUserId);
-    expect(res2.result[res2.result.length-1].created_at).to.not.equal(null);
-    expect(res2.result[res2.result.length-1].deleted_at).to.equal(null);
-    expect(res2.result[res2.result.length-1].is_deleted).to.equal(false);
+    expect(res2.result[res2.result.length - 1].name).to.equal('Laser Class Testy');
+    expect(res2.result[res2.result.length - 1].created_by).to.equal(myUserId);
+    expect(res2.result[res2.result.length - 1].created_at).to.not.equal(null);
+    expect(res2.result[res2.result.length - 1].deleted_at).to.equal(null);
+    expect(res2.result[res2.result.length - 1].is_deleted).to.equal(false);
   });
 
   lab.test('should error with missing name', async () => {
diff --git a/test/routes/events/browse.js b/test/routes/events/browse.js
index 3530e65..cb82317 100644
--- a/test/routes/events/browse.js
+++ b/test/routes/events/browse.js
@@ -18,8 +18,9 @@ const lab = exports.lab = require('lab').script();
 const url = require('url');
 
 const server = require('../../..');
-const { createMapRelations, destroyRecords, getAuthToken } = require('../../fixture-client');
+const { getAuthToken } = require('../../fixture-client');
 const knex = require('../../../knex');
+const clearDb = require('../../clearDb');
 
 const { users, events } = require('../../fixtures');
 
@@ -34,7 +35,7 @@ lab.experiment('GET /events/', () => {
   });
 
   lab.after(async () => {
-    await destroyRecords({ users, events });
+    await clearDb();
   });
 
   lab.test('should retrieve event information when logged in', (done) => {
diff --git a/test/routes/events/delete.js b/test/routes/events/delete.js
index ab966aa..6b8f90a 100644
--- a/test/routes/events/delete.js
+++ b/test/routes/events/delete.js
@@ -19,19 +19,18 @@ const url = require('url');
 
 const server = require('../../..');
 const knex = require('../../../knex');
+const clearDb = require('../../clearDb');
 
-const { createMapRelations, destroyRecords, getAuthToken, fixtures, makeUserIdAdmin } = require('../../fixture-client');
-const { events, users } = require('../../fixtures');
+const { getAuthToken, makeUserIdAdmin } = require('../../fixture-client');
+const { users } = require('../../fixtures');
 
 lab.experiment('DELETE /events/', () => {
   let Authorization;
   let myUserId;
-  let memberships;
-  let events;
 
   lab.before(async () => {
-    let insertedUserIds = await knex('users').insert(users).returning(['id']);
-    myUserId = insertedUserIds[0]['id'];
+    const insertedUserIds = await knex('users').insert(users).returning(['id']);
+    myUserId = insertedUserIds[0].id;
 
     await makeUserIdAdmin(myUserId);
 
@@ -39,8 +38,8 @@ lab.experiment('DELETE /events/', () => {
     Authorization = authRes.token;
   });
 
-  lab.after(() => {
-    return destroyRecords({ memberships }).then(destroyRecords({ events })).then(destroyRecords({ users }));
+  lab.after(async () => {
+    await clearDb();
   });
 
   lab.test('should successfully delete an event', async () => {
diff --git a/test/routes/events/edit.js b/test/routes/events/edit.js
index 6f543de..d8b9418 100644
--- a/test/routes/events/edit.js
+++ b/test/routes/events/edit.js
@@ -19,16 +19,17 @@ const url = require('url');
 const { omit } = require('lodash');
 
 const server = require('../../..');
-const { createMapRelations, destroyRecords, getAuthToken, fixtures, makeUserIdAdmin } = require('../../fixture-client');
+const { getAuthToken, makeUserIdAdmin } = require('../../fixture-client');
 const { users, events } = require('../../fixtures');
 const knex = require('../../../knex');
+const clearDb = require('../../clearDb');
 
 lab.experiment('PUT /events/', () => {
   let Authorization;
 
   lab.before(async () => {
-    let insertedUserIds = await knex('users').insert(users).returning(['id']);
-    myUserId = insertedUserIds[0]['id'];
+    const insertedUserIds = await knex('users').insert(users).returning(['id']);
+    const myUserId = insertedUserIds[0].id;
     await knex('events').insert(events);
 
     await makeUserIdAdmin(myUserId);
@@ -38,13 +39,12 @@ lab.experiment('PUT /events/', () => {
   });
 
   lab.after(async () => {
-    return destroyRecords({ memberships }).then(destroyRecords({ events })).then(destroyRecords({ users }));
+    await clearDb();
   });
 
   lab.test('should successfully edit an event', async () => {
     const eventsToEdit = await knex('events');
-    console.log(eventsToEdit);
-    eventsToEdit[0].name = "fookie";
+    eventsToEdit[0].name = 'fookie';
 
     const options = {
       url: url.format(`/events/${eventsToEdit[0].id}`),
@@ -54,7 +54,6 @@ lab.experiment('PUT /events/', () => {
     };
 
     const res = await server.inject(options);
-    console.log(res);
     expect(res.statusCode).to.equal(200);
     expect(res.result.id).to.equal(eventsToEdit[0].id);
     expect(res.result.name).to.equal('fookie');
diff --git a/test/routes/events/read.js b/test/routes/events/read.js
index 6512897..55ca07a 100644
--- a/test/routes/events/read.js
+++ b/test/routes/events/read.js
@@ -18,10 +18,10 @@ const lab = exports.lab = require('lab').script();
 const url = require('url');
 
 const server = require('../../..');
-const { createMapRelations, destroyRecords, getAuthToken, fixtures } = require('../../fixture-client');
+const { getAuthToken } = require('../../fixture-client');
 const knex = require('../../../knex');
-
 const { users, events } = require('../../fixtures');
+const clearDb = require('../../clearDb');
 
 lab.experiment('GET /events/{event_id}', () => {
   let Authorization;
@@ -34,7 +34,7 @@ lab.experiment('GET /events/{event_id}', () => {
   });
 
   lab.after(async () => {
-    await destroyRecords({ users, events });
+    await clearDb();
   });
 
   lab.test('should return a event by id', async () => {
@@ -56,7 +56,7 @@ lab.experiment('GET /events/{event_id}', () => {
   lab.test.skip('should error if event is not found', async () => {
     const event = await knex('events').offset(0).first('id');
     const options = {
-      url: url.format(`/events/badbadf7-53a7-4d66-abf5-541d3ed767d0`),
+      url: url.format('/events/badbadf7-53a7-4d66-abf5-541d3ed767d0'),
       method: 'GET',
       headers: { Authorization },
     };
@@ -68,7 +68,7 @@ lab.experiment('GET /events/{event_id}', () => {
   lab.test('should error with bad event id', async () => {
     const event = await knex('events').offset(0).first('id');
     const options = {
-      url: url.format(`/events/badbadbad`),
+      url: url.format('/events/badbadbad'),
       method: 'GET',
       headers: { Authorization },
     };
diff --git a/test/routes/users/add.js b/test/routes/users/add.js
index cfc5317..17ec9a9 100644
--- a/test/routes/users/add.js
+++ b/test/routes/users/add.js
@@ -21,16 +21,15 @@ const { omit } = require('lodash');
 
 const bread = require('../../../lib/bread');
 const server = require('../../..');
-const { destroyRecords, destroyTokens } = require('../../fixture-client');
 const { users } = require('../../fixtures');
+const clearDb = require('../../clearDb');
 
 lab.experiment('POST /user', () => {
   // eslint-disable-next-line
   let tokens = [];
 
-  lab.after(() => {
-    return destroyRecords({ users })
-      .then(destroyTokens(tokens));
+  lab.after(async () => {
+    await clearDb();
   });
 
   lab.test('should create a user', async () => {
diff --git a/test/routes/users/browse.js b/test/routes/users/browse.js
index decacb3..90a6c5f 100644
--- a/test/routes/users/browse.js
+++ b/test/routes/users/browse.js
@@ -18,9 +18,10 @@ const lab = exports.lab = require('lab').script();
 const url = require('url');
 
 const server = require('../../..');
-const { destroyRecords, getAuthToken, fixtures } = require('../../fixture-client');
+const { getAuthToken } = require('../../fixture-client');
 const { users } = require('../../fixtures');
 const knex = require('../../../knex');
+const clearDb = require('../../clearDb');
 
 lab.experiment('GET /users/', () => {
   let user;
@@ -32,8 +33,8 @@ lab.experiment('GET /users/', () => {
     Authorization = authRes.token;
   });
 
-  lab.after(() => {
-    return destroyRecords({ users });
+  lab.after(async () => {
+    await clearDb();
   });
 
   lab.test('should retrieve my information when logged in', (done) => {
diff --git a/test/routes/users/delete.js b/test/routes/users/delete.js
index 784d428..cfc1b66 100644
--- a/test/routes/users/delete.js
+++ b/test/routes/users/delete.js
@@ -18,9 +18,10 @@ const lab = exports.lab = require('lab').script();
 const url = require('url');
 
 const server = require('../../..');
-const { destroyRecords, getAuthToken, fixtures } = require('../../fixture-client');
+const { getAuthToken } = require('../../fixture-client');
 const { users } = require('../../fixtures');
 const knex = require('../../../knex');
+const clearDb = require('../../clearDb');
 
 lab.experiment('DELETE /users/', () => {
   let Authorization;
@@ -31,8 +32,8 @@ lab.experiment('DELETE /users/', () => {
     Authorization = authRes.token;
   });
 
-  lab.after(() => {
-    return destroyRecords({ users });
+  lab.after(async () => {
+    await clearDb();
   });
 
   lab.test('should fail if trying to delete an unauthorized user', async () => {
diff --git a/test/routes/users/read.js b/test/routes/users/read.js
index 371ec71..1ad41f8 100644
--- a/test/routes/users/read.js
+++ b/test/routes/users/read.js
@@ -18,9 +18,10 @@ const lab = exports.lab = require('lab').script();
 const url = require('url');
 
 const server = require('../../..');
-const { destroyRecords, getAuthToken, fixtures } = require('../../fixture-client');
+const { getAuthToken } = require('../../fixture-client');
 const { users } = require('../../fixtures');
 const knex = require('../../../knex');
+const clearDb = require('../../clearDb');
 
 lab.experiment('GET /users/{user_id}', () => {
   let Authorization;
@@ -31,8 +32,8 @@ lab.experiment('GET /users/{user_id}', () => {
     Authorization = authRes.token;
   });
 
-  lab.after(() => {
-    return destroyRecords({ users });
+  lab.after(async () => {
+    await clearDb();
   });
 
   lab.test('should return a user by id', async () => {
```

image

**Comment by zyphlar** [comment]
@duncanbeevers nice thanks! I was wondering how/where to call clearDb. Love them green checks!

**Comment by duncanbeevers** [comment]
### TL;DR

No action necessary on your part for this comment; just gathering some info.

### Lab onCleanup

lab also supports an `onCleanup` hook that can be provided to tests, which could make tests a little better citizens in cleaning up after themselves.
https://hapi.dev/module/lab/api/?v=25.0.1#oncleanup

image

### Knex Transactions

It looks like it's possible to use postgres transaction support to run tests in a more isolated fashion, but I may have to thread the knex transaction instance into the request handler, which sounds like it might couple the code together a bit.

I think the ideal solution would be to begin a transaction at the beginning of each test, and to roll it back at the end of the test, but I was having some difficulty in getting that setup.
